cmd/cloudwatcher/metrics.go

```diff
@@ -0,0 +1,506 @@
+
+	s.vcpuCounts = append(s.vcpuCounts, float64(vcpuCount)) // Add vCPU count.
+
+	// Add 1m Load.
+	load, ok := node.Stats.OS.CPU.LoadAverage["1m"]
+	if !ok {
+		panic("missing 1m load average")
+	}
+	s.load1m = append(s.load1m, load)
+
+	// Add 5m Load.
+	load, ok = node.Stats.OS.CPU.LoadAverage["5m"]
+	if !ok {
+		panic("missing 5m load average")
+	}
+	s.load5m = append(s.load5m, load)
+
+	// Add 15m Load.
+	load, ok = node.Stats.OS.CPU.LoadAverage["15m"]
+	if !ok {
+		panic("missing 15m load average")
+	}
+	s.load15m = append(s.load15m, load)
+
+	// Add JVM heap totals
+	s.maxJVMHeapBytes = append(s.maxJVMHeapBytes, float64(node.Stats.JVM.Mem.HeapMaxInBytes))
+	s.usedJVMHeapBytes = append(s.usedJVMHeapBytes, float64(node.Stats.JVM.Mem.HeapUsedInBytes))
+
+	// Add JVM heap stats for each memory pool
+	for name, p := range node.Stats.JVM.Mem.Pools {
+		d, ok := s.jvmHeapPools[name]
+		if !ok {
+			d = &struct {
+				Used     []float64
+				Max      []float64
+				PeakUsed []float64
+				PeakMax  []float64
+			}{
+				Used:     make([]float64, 0),
+				Max:      make([]float64, 0),
+				PeakUsed: make([]float64, 0),
+				PeakMax:  make([]float64, 0),
+			}
+			s.jvmHeapPools[name] = d
+		}
+		d.Used = append(d.Used, float64(p.UsedInBytes))
+		d.Max = append(d.Max, float64(p.MaxInBytes))
+		d.PeakUsed = append(d.PeakUsed, float64(p.PeakUsedInBytes))
+		d.PeakMax = append(d.PeakMax, float64(p.PeakMaxInBytes))
+	}
+
+	// Add JVM garbage collection stats.
+	for name, c := range node.Stats.JVM.GC.Collectors {
+		d, ok := s.garbageCollectors[name]
+		if !ok {
+			d = &struct {
+				Count []float64
+				Time  []float64
+			}{
+				Count: make([]float64, 0),
+				Time:  make([]float64, 0),
+			}
+			s.garbageCollectors[name] = d
+		}
+		d.Count = append(d.Count, float64(c.CollectionCount))
+		d.Time = append(d.Time, float64(c.CollectionTimeInMillis))
+	}
+
+	// Add filesystem stats for data nodes.
+	if str.In("data", node.Roles...) {
+		s.maxFSBytes = append(s.maxFSBytes, float64(node.Stats.FS.Total.TotalInBytes))
+		s.availableFSBytes = append(s.availableFSBytes, float64(node.Stats.FS.Total.AvailableInBytes))
+		s.excludedFromAllocation = append(s.excludedFromAllocation, node.ExcludedShardAllocation)
+	} else {
+		// For non-data nodes, append zeros just to keep slice sizes consistent.
+		s.maxFSBytes = append(s.maxFSBytes, 0)
+		s.availableFSBytes = append(s.availableFSBytes, 0)
+		s.excludedFromAllocation = append(s.excludedFromAllocation, false)
+	}
+}
+
+// Metrics returns the CloudWatch metric data points.
+func (s *metricsCollector) Metrics(dimensions []*cloudwatch.Dimension, timestamp time.Time) []*cloudwatch.MetricDatum {
+	metrics := make([]*cloudwatch.MetricDatum, 0) // The output slice of CloudWatch data points.
+
+	if s.count == 0 { // Shortcut for no nodes Add()-ed.
+		return metrics
+	}
+
+	// Sum vCPU counts.
+	vcpuCount := sum(s.vcpuCounts...)
+
+	// Sum loads.
+	load1m := sum(s.load1m...)
```

Comment:
Why do you use variatic args instead of passing a slice? Is that just how sum() works?